File: EticaretJava/app/src/main/java/com/example/eticaretjava/viewmodel/ProductDetailViewModel.java
package com.example.eticaretjava.viewmodel;


import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.eticaretjava.model.Product.ProductDto;
import com.example.eticaretjava.repo.Repositories;
import com.example.eticaretjava.repo.ResultCallback;

public class ProductDetailViewModel extends ViewModel {

    public static class DetailState {
        public boolean loading = false;
        public String error = null;
        public ProductDto product = null;
        public boolean addSuccess = false;
    }

    private final MutableLiveData<DetailState> state = new MutableLiveData<>(new DetailState());

    private final Repositories.ProductRepository productRepo;
    private final Repositories.CartRepository cartRepo;

    public ProductDetailViewModel(Repositories.ProductRepository productRepo, Repositories.CartRepository cartRepo) {
        this.productRepo = productRepo;
        this.cartRepo = cartRepo;
    }

    public LiveData<DetailState> getState() {
        return state;
    }

    private DetailState copyState() {
        DetailState cur = state.getValue() != null ? state.getValue() : new DetailState();
        DetailState n = new DetailState();
        n.loading = cur.loading;
        n.error = cur.error;
        n.product = cur.product;
        n.addSuccess = cur.addSuccess;
        return n;
    }

    public void load(int id) {
        DetailState s = copyState();
        s.loading = true;
        s.error = null;
        s.addSuccess = false;
        state.setValue(s);

        productRepo.getProduct(id, new ResultCallback<ProductDto>() {
            @Override
            public void onSuccess(ProductDto data) {
                DetailState ns = copyState();
                ns.loading = false;
                ns.product = data;
                state.postValue(ns);
            }

            @Override
            public void onError(String message) {
                DetailState ns = copyState();
                ns.loading = false;
                ns.error = message != null ? message : "Detay hatası";
                state.postValue(ns);
            }
        });
    }

    public void addToCart(int productId, int qty) {
        DetailState s = copyState();
        s.loading = true;
        s.error = null;
        s.addSuccess = false;
        state.setValue(s);

        cartRepo.addToCart(productId, qty, new ResultCallback<com.example.eticaretkotlin.model.Cart.AddToCartResponse>() {
            @Override
            public void onSuccess(com.example.eticaretkotlin.model.Cart.AddToCartResponse data) {
                DetailState ns = copyState();
                ns.loading = false;
                ns.addSuccess = true;
                state.postValue(ns);
            }

            @Override
            public void onError(String message) {
                DetailState ns = copyState();
                ns.loading = false;
                ns.error = message != null ? message : "Sepete ekleme hatası";
                state.postValue(ns);
            }
        });
    }

    public void clearFlags() {
        DetailState ns = copyState();
        ns.addSuccess = false;
        ns.error = null;
        state.setValue(ns);
    }
}

File: EticaretKotlin/app/src/main/java/com/example/eticaretkotlin/viewmodel/ProductDetailViewModel.kt
package com.example.eticaretkotlin.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.eticaretkotlin.model.ProductDto
import com.example.eticaretkotlin.repo.CartRepository
import com.example.eticaretkotlin.repo.ProductRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class ProductDetailViewModel(
    private val productRepo: ProductRepository,
    private val cartRepo: CartRepository
) : ViewModel() {

    data class DetailState(
        val loading: Boolean = false,
        val error: String? = null,
        val product: ProductDto? = null,
        val addSuccess: Boolean = false
    )

    private val _state = MutableStateFlow(DetailState())
    val state: StateFlow<DetailState> = _state

    fun load(id: Int) = viewModelScope.launch {
        _state.update { it.copy(loading = true, error = null, addSuccess = false) }

        productRepo.getProduct(id)
            .onSuccess { prod ->
                _state.update { it.copy(loading = false, product = prod) }
            }
            .onFailure { e ->
                _state.update { it.copy(loading = false, error = e.message ?: "Detay hatası") }
            }
    }

    fun addToCart(productId: Int, qty: Int) = viewModelScope.launch {
        _state.update { it.copy(loading = true, error = null, addSuccess = false) }

        cartRepo.addToCart(productId, qty)
            .onSuccess {
                _state.update { it.copy(loading = false, addSuccess = true) }
            }
            .onFailure { e ->
                _state.update { it.copy(loading = false, error = e.message ?: "Sepete ekleme hatası") }
            }
    }

    fun clearFlags() = _state.update { it.copy(addSuccess = false, error = null) }
}
